File: lib/presets/matribox_family_expansion_plan.dart
/// `FAMILY_EXPANSION_P01_V1`: the single, most informative next hardware
/// test. [FamilyExpansionPlan] is the OFFLINE data of the plan (steps with
/// their evidence purpose; validated against the evidence model in
/// test/matribox_evidence_v2_test.dart). [FamilyExpansionP01Plan] is the
/// debug/certification wrapper that runs those very operations through the
/// existing Full Live session/executor/verifier (see
/// matribox_family_expansion_certification.dart; own compile gate, only the
/// plan id and the backup hash cross the channel, the native side holds its
/// own constant operation list). This file stays pure data.
///
/// Rules of the plan: User P01 only; fresh read + backup + hash first; every
/// operation is catalog/capture-resolved (no free ids/indices); values are
/// values the editor itself wrote in the big capture (no extremes, no random
/// values); no Store, no metadata, no name/BPM/VOL. After the live run:
/// physical check -> manual device save -> fresh readback.
///
/// Nothing here releases a productive family.
library;

import 'matribox_chain_slot.dart';
import 'matribox_full_live_plan.dart';
import 'matribox_model_library.dart';

class FamilyExpansionStep {
  const FamilyExpansionStep(this.operation, this.purpose, {this.optional = false});
  final FullLiveOperation operation;

  /// The evidence question this operation answers.
  final String purpose;

  /// Tier B: closes a family gap that is not needed for song no. 2 itself.
  final bool optional;
}

class FamilyExpansionPlan {
  FamilyExpansionPlan(MatriboxModelLibrary library) : steps = _build(library);

  static const planId = 'FAMILY_EXPANSION_P01_V1';
  final List<FamilyExpansionStep> steps;

  List<FullLiveOperation> operations({bool includeOptional = false}) => [
    for (final s in steps)
      if (includeOptional || !s.optional) s.operation,
  ];

  static List<FamilyExpansionStep> _build(MatriboxModelLibrary library) {
    FullLiveOperation select(MatriboxChainSlot slot, String name) =>
        FullLiveOperation.model(slot, library.byName(slot, name)!.algorithm);
    FullLiveOperation param(MatriboxChainSlot slot, String model, String parameter, double value) =>
        FullLiveOperation.parameter(slot, library.byName(slot, model)!.algorithm, parameter, value);
    FullLiveOperation on(MatriboxChainSlot slot) => FullLiveOperation.toggle(slot, enabled: true);
    const fx2 = MatriboxChainSlot.fx2, nr = MatriboxChainSlot.nr, cab = MatriboxChainSlot.cab;
    const eq = MatriboxChainSlot.eq, mod = MatriboxChainSlot.mod, dly = MatriboxChainSlot.dly;
    const rvb = MatriboxChainSlot.rvb;
    return List.unmodifiable([
      FamilyExpansionStep(select(fx2, 'Boost'), 'Boost-Auswahl als sauberer Ausgangspunkt für den Drive-Parameter (exakt bereits bestätigt).'),
      FamilyExpansionStep(param(fx2, 'Boost', 'Gain', 23), 'Drive: erster Parameterwrite auf einem nur-Katalog-Modell in FX2 (Zahl, Index 0 wie Skreamer/Blues OD).'),
      FamilyExpansionStep(param(fx2, 'Boost', 'Bright', 1), 'Boolescher Switch 0/1 in FX2 (dritter Slot mit Flag-Sample neben DLY und RVB: gibt der Flag-Familie slotübergreifende Breite).'),
      FamilyExpansionStep(select(nr, 'Gate 2'), 'MODEL SELECT in NR (neue Kategorie, capture-bestätigter Code).'),
      FamilyExpansionStep(param(nr, 'Gate 2', 'THRE', 31), 'Gate2 THRE (Zahl in NR; Wert aus dem Capture) für die Gate-Heuristik.'),
      FamilyExpansionStep(on(nr), 'Block-CC ON auf NR (bisher nie aktiv gesendet).'),
      FamilyExpansionStep(select(cab, 'Sol 4x12'), 'Sol 4x12 als Ausgangspunkt für CAB VOL (exakt bereits bestätigt).'),
      FamilyExpansionStep(param(cab, 'Sol 4x12', 'VOL', 43), 'CAB VOL: Wire-Index 1 (Hersteller-ID 2 − 1) auf einem nur-Katalog-CAB (bisher nur BritGN im Capture).'),
      FamilyExpansionStep(select(eq, 'Guitar EQ'), 'MODEL SELECT nur-Katalog-Modell in EQ (neue Kategorie; Guitar EQ ist der Recipe-Standard).'),
      FamilyExpansionStep(param(eq, 'Guitar EQ', '400Hz', -23), 'Negativer Float32 (-23 stammt aus dem Capture) auf einem nur-Katalog-Parameter in EQ.'),
      FamilyExpansionStep(on(eq), 'Block-CC ON auf EQ (bisher nur OFF bestätigt).'),
      FamilyExpansionStep(select(mod, 'Chorus A'), 'MODEL SELECT nur-Katalog-Modell in MOD.'),
      FamilyExpansionStep(param(mod, 'Chorus A', 'Rate', 3.7), 'Dezimaler Float32 (3.7 aus dem Capture, Katalog-Step 0.1); Sync bleibt unberührt.'),
      FamilyExpansionStep(select(dly, 'Warm'), 'MODEL SELECT nur-Katalog-Modell in DLY.'),
      FamilyExpansionStep(param(dly, 'Warm', 'Time', 743), 'Zeitparameter 20..4000 ms (Wert aus dem Sweep-Capture).'),
      FamilyExpansionStep(param(dly, 'Warm', 'Trail', 1), 'Boolescher Switch 0/1 in DLY auf einem nur-Katalog-Parameter.'),
      FamilyExpansionStep(select(rvb, 'Room'), 'MODEL SELECT nur-Katalog-Modell in RVB (Recipe-Standard für wenig Hall).'),
      FamilyExpansionStep(param(rvb, 'Room', 'Mix', 23), 'Reverb Mix (Zahl, Wert aus Mod RVB) für die Reverb-Normalisierung.'),
      FamilyExpansionStep(param(rvb, 'Room', 'Decay', 41), 'Reverb Decay (Zahl, Wert aus Mod RVB).'),
      FamilyExpansionStep(param(rvb, 'Room', 'Trail', 1), 'Boolescher Switch 0/1 in RVB.'),
      FamilyExpansionStep(on(rvb), 'Block-CC ON auf RVB (Controller 0x38, äußerster Slot; bisher nie aktiv).'),
      FamilyExpansionStep(select(MatriboxChainSlot.fx1, 'Boost'), 'Tier B: MODEL SELECT nur-Katalog-Modell in FX1 (Zwilling von FX2).', optional: true),
      FamilyExpansionStep(select(MatriboxChainSlot.amp, 'Sol 100 OD'), 'Tier B: MODEL SELECT in AMP (Code aus Geräte-Reads); schließt die AMP-Familie.', optional: true),
    ]);
  }
}

File: android/app/src/main/kotlin/de/neevel/wyrmtone/MatriboxFamilyExpansionPlan.kt
package de.neevel.wyrmtone

/**
 * The fixed `FAMILY_EXPANSION_P01_V1` certification plan for User P01: one
 * controlled run that actively tests the model-select, number, signed,
 * decimal, flag and block-CC families still missing hardware evidence
 * (Tier B, FX1 Boost and AMP Sol 100 OD select, is part of it). It runs
 * through the existing Full Live session, port and codec (no second
 * transport). Dart addresses it only by [PLAN_ID] plus the verified backup
 * hash; no operation, byte, slot, algorithm id or parameter index ever comes
 * from the caller. No Store, no `12 11`, no name/BPM/VOL, no User IR.
 *
 * Sent in chain order: slots ascend FX1..RVB; inside a slot MODEL SELECT,
 * then PARAMETERs, then the block CC (a select resets the block's
 * parameters, and a block is switched ON only once it is configured). Must
 * stay identical to lib/presets/matribox_family_expansion_certification.dart
 * (golden test).
 *
 * The run was CERTIFIED on real hardware (2026-09-21, wire = manufacturer ID - 1).
 * The productive families follow from its samples through the generated
 * manufacturer table ([MatriboxToneTransferCatalog]); this certification plan
 * and its own table stay independent of the productive path.
 */
object MatriboxFamilyExpansionPlan {
    const val PLAN_ID = "FAMILY_EXPANSION_P01_V1"

    private const val BOOST = 0x0000001a
    private const val SOL_100_OD = 0x07000047
    private const val GATE_2 = 0x0000001d
    private const val SOL_4X12 = 0x0a000028
    private const val GUITAR_EQ = 0x01000035
    private const val CHORUS_A = 0x04000000
    private const val WARM = 0x0b00000d
    private const val ROOM = 0x0c000000

    val operations: List<FullLiveOperation> = listOf(
        // Tier B
        FullLiveOperation.ModelSelect(ChainSlot.FX1, BOOST, "Boost"),
        FullLiveOperation.ModelSelect(ChainSlot.FX2, BOOST, "Boost"),
        FullLiveOperation.Parameter(ChainSlot.FX2, BOOST, 0, 23f, "Gain"),
        // Bright: editor ID 3, wire index 2 (ID 2 is a hidden parameter); wire = ID - 1
        FullLiveOperation.Parameter(ChainSlot.FX2, BOOST, 2, 1f, "Bright"),
        // Tier B
        FullLiveOperation.ModelSelect(ChainSlot.AMP, SOL_100_OD, "Sol 100 OD"),
        FullLiveOperation.ModelSelect(ChainSlot.NR, GATE_2, "Gate 2"),
        FullLiveOperation.Parameter(ChainSlot.NR, GATE_2, 0, 31f, "THRE"),
        FullLiveOperation.BlockToggle(ChainSlot.NR, enabled = true),
        FullLiveOperation.ModelSelect(ChainSlot.CAB, SOL_4X12, "Sol 4x12"),
        // CAB VOL: editor ID 2, wire index 1 (wire = ID - 1)
        FullLiveOperation.Parameter(ChainSlot.CAB, SOL_4X12, 1, 43f, "VOL"),
        FullLiveOperation.ModelSelect(ChainSlot.EQ, GUITAR_EQ, "Guitar EQ"),
        FullLiveOperation.Parameter(ChainSlot.EQ, GUITAR_EQ, 1, -23f, "400Hz"),
        FullLiveOperation.BlockToggle(ChainSlot.EQ, enabled = true),
        FullLiveOperation.ModelSelect(ChainSlot.MOD, CHORUS_A, "Chorus A"),
        FullLiveOperation.Parameter(ChainSlot.MOD, CHORUS_A, 1, 3.7f, "Rate"),
        FullLiveOperation.ModelSelect(ChainSlot.DLY, WARM, "Warm"),
        FullLiveOperation.Parameter(ChainSlot.DLY, WARM, 1, 743f, "Time"),
        FullLiveOperation.Parameter(ChainSlot.DLY, WARM, 4, 1f, "Trail"),
        FullLiveOperation.ModelSelect(ChainSlot.RVB, ROOM, "Room"),
        FullLiveOperation.Parameter(ChainSlot.RVB, ROOM, 0, 23f, "Mix"),
        FullLiveOperation.Parameter(ChainSlot.RVB, ROOM, 2, 41f, "Decay"),
        FullLiveOperation.Parameter(ChainSlot.RVB, ROOM, 3, 1f, "Trail"),
        FullLiveOperation.BlockToggle(ChainSlot.RVB, enabled = true),
    )
}

/** What a certification run is bound to: User P01 and the verified BEFORE-backup hash. */
internal data class CertificationTarget(val bank: String, val slot: Int, val backupHash: String)

/**
 * Independent native trust table for the family-expansion plan: the
 * manufacturer-catalog identities (slot/category, algorithm code, parameter
 * name, wire index, type and range) of exactly the models the plan selects.
 * The plan is validated against this table BEFORE the port is opened; one
 * mismatch rejects every operation (ZERO SENDS). Nothing here is hardware
 * evidence, and nothing here makes another model or parameter sendable.
 */
internal object MatriboxFamilyExpansionCatalog {
    enum class Kind { NUMBER, SIGNED, DECIMAL, FLAG }

    data class Param(val name: String, val wireIndex: Int, val kind: Kind, val min: Float, val max: Float)

    data class Model(val name: String, val code: Int, val slots: Set<ChainSlot>, val parameters: List<Param>)

    private fun knob(name: String, index: Int, max: Float = 99f) = Param(name, index, Kind.NUMBER, 0f, max)
    private fun eqBand(name: String, index: Int) = Param(name, index, Kind.SIGNED, -50f, 50f)
    private fun flag(name: String, index: Int) = Param(name, index, Kind.FLAG, 0f, 1f)

    val models: List<Model> = listOf(
        // wire index = editor ID - 1: Bright has ID 3 (ID 2 is hidden) -> wire 2
        Model("Boost", 0x0000001a, setOf(ChainSlot.FX1, ChainSlot.FX2), listOf(knob("Gain", 0), flag("Bright", 2))),
        Model(
            "Sol 100 OD", 0x07000047, setOf(ChainSlot.AMP),
            listOf(knob("Gain", 0), knob("PRES", 1), knob("Master", 2), knob("Bass", 3), knob("Middle", 4), knob("Treble", 5)),
        ),
        Model("Gate 2", 0x0000001d, setOf(ChainSlot.NR), listOf(knob("THRE", 0), knob("ATK", 1), knob("Rel", 2))),
        // CAB VOL: editor ID 2 -> wire index 1 (wire = ID - 1).
        Model("Sol 4x12", 0x0a000028, setOf(ChainSlot.CAB), listOf(knob("VOL", 1))),
        Model(
            "Guitar EQ", 0x01000035, setOf(ChainSlot.EQ),
            listOf(eqBand("125Hz", 0), eqBand("400Hz", 1), eqBand("800Hz", 2), eqBand("1.6kHz", 3), eqBand("4kHz", 4), knob("VOL", 5)),
        ),
        Model(
            "Chorus A", 0x04000000, setOf(ChainSlot.MOD),
            listOf(knob("Depth", 0), Param("Rate", 1, Kind.DECIMAL, 0.1f, 10f), knob("Tone", 2), flag("Sync", 3)),
        ),
        Model(
            "Warm", 0x0b00000d, setOf(ChainSlot.DLY),
            listOf(knob("Mix", 0), knob("Time", 1, 4000f).copy(min = 20f), knob("FdBk", 2), flag("Sync", 3), flag("Trail", 4)),
        ),
        Model("Room", 0x0c000000, setOf(ChainSlot.RVB), listOf(knob("Mix", 0), knob("Pre Delay", 1, 100f), knob("Decay", 2), flag("Trail", 3))),
    )


    fun modelByCode(code: Int): Model? = models.firstOrNull { it.code == code }
}

/**
 * Whole-plan preflight of `FAMILY_EXPANSION_P01_V1`, run before the port is
 * opened. It is deliberately independent of the plan constant (it validates
 * whatever list it is given), so an injected invalid list is provably
 * rejected with zero sends.
 */
internal object MatriboxFamilyExpansionValidator {
    private const val MAX_OPERATIONS = 64
    private val HASH = Regex("^[0-9a-f]{64}$")

    private fun reject(code: String, message: String): Nothing = throw ToneTransferRejection(code, message)

    private fun slotOf(op: FullLiveOperation) = when (op) {
        is FullLiveOperation.ModelSelect -> op.slot
        is FullLiveOperation.Parameter -> op.slot
        is FullLiveOperation.BlockToggle -> op.slot
    }

    private fun rank(op: FullLiveOperation) = when (op) {
        is FullLiveOperation.ModelSelect -> 0
        is FullLiveOperation.Parameter -> 1
        is FullLiveOperation.BlockToggle -> 2
    }

    /** Returns null when the plan is not this test's, or the rejection message when it is invalid. */
    fun preflightMessage(planId: String, operations: List<FullLiveOperation>, target: CertificationTarget?): String? {
        if (planId != MatriboxFamilyExpansionPlan.PLAN_ID) return null
        return try {
            validate(planId, operations, target)
            null
        } catch (rejection: ToneTransferRejection) {
            "${rejection.code}: ${rejection.message}"
        }
    }

    fun validate(planId: String, operations: List<FullLiveOperation>, target: CertificationTarget?) {
        if (planId != MatriboxFamilyExpansionPlan.PLAN_ID) reject("PLAN_ID", "Falsche Plan-ID.")
        if (target == null) reject("TARGET_MISSING", "Ziel (User P01) und Backup-Hash fehlen.")
        if (target.bank != "USER") reject("TARGET_NOT_ALLOWED", "Nur die User-Bank ist freigegeben.")
        if (target.slot != 1) reject("TARGET_NOT_ALLOWED", "Nur User P01 ist freigegeben.")
        if (!HASH.matches(target.backupHash)) reject("BACKUP_HASH_INVALID", "Backup-Hash muss ein SHA-256 (hex) sein.")
        if (operations.isEmpty()) reject("EMPTY_PLAN", "Der Plan enthält keine Operation.")
        if (operations.size > MAX_OPERATIONS) reject("PLAN_TOO_LARGE", "Zu viele Operationen.")

        val selectedInSlot = mutableMapOf<ChainSlot, Int>()
        val written = mutableSetOf<Pair<ChainSlot, String>>()
        val toggled = mutableSetOf<ChainSlot>()
        var lastSlot = -1
        var lastRank = -1
        for ((index, op) in operations.withIndex()) {
            val slot = slotOf(op)
            val r = rank(op)
            if (operations.take(index).any { it == op }) reject("DUPLICATE_OPERATION", "Operation #$index ist doppelt.")
            // deterministic chain order: slots ascend; SELECT, then PARAMETERs, then the block CC
            if (slot.ordinal < lastSlot || (slot.ordinal == lastSlot && r < lastRank)) {
                reject("ORDER_INVALID", "Operation #$index (${op.label}) steht in falscher Reihenfolge.")
            }
            if (slot.ordinal != lastSlot) lastRank = -1
            lastSlot = slot.ordinal
            lastRank = r

            when (op) {
                is FullLiveOperation.ModelSelect -> {
                    val model = MatriboxFamilyExpansionCatalog.modelByCode(op.code)
                        ?: reject("MODEL_UNKNOWN", "Operation #$index: Algorithmus 0x${op.code.toString(16)} ist nicht im Katalog.")
                    if (model.name != op.name) reject("MODEL_UNKNOWN", "Operation #$index: Name/Code passen nicht (${op.name}).")
                    if (slot !in model.slots) reject("MODEL_WRONG_SLOT", "Operation #$index: ${model.name} gehört nicht zu ${slot.name}.")
                    if (slot in selectedInSlot) reject("CONFLICTING_OPERATIONS", "Operation #$index: ${slot.name} wählt ein zweites Modell.")
                    selectedInSlot[slot] = op.code
                }
                is FullLiveOperation.Parameter -> {
                    val model = MatriboxFamilyExpansionCatalog.modelByCode(op.code)
                        ?: reject("MODEL_UNKNOWN", "Operation #$index: Algorithmus 0x${op.code.toString(16)} ist nicht im Katalog.")
                    if (slot !in model.slots) reject("MODEL_WRONG_SLOT", "Operation #$index: ${model.name} gehört nicht zu ${slot.name}.")
                    if (selectedInSlot[slot] != op.code) {
                        reject("PARAMETER_MODEL_NOT_SELECTED", "Operation #$index: ${model.name} ist in ${slot.name} nicht gewählt.")
                    }
                    val parameter = model.parameters.firstOrNull { it.name == op.name }
                        ?: reject("PARAMETER_UNKNOWN", "Operation #$index: ${model.name}/${op.name} ist nicht im Katalog.")
                    if (parameter.wireIndex != op.index) {
                        reject("PARAMETER_INDEX_MISMATCH", "Operation #$index: Wire-Index ${op.index} statt ${parameter.wireIndex}.")
                    }
                    if (!op.value.isFinite() || op.value < parameter.min || op.value > parameter.max) {
                        reject("VALUE_OUT_OF_RANGE", "Operation #$index: ${op.value} außerhalb ${parameter.min}..${parameter.max}.")
                    }
                    when (parameter.kind) {
                        MatriboxFamilyExpansionCatalog.Kind.FLAG ->
                            if (op.value != 0f && op.value != 1f) reject("VALUE_TYPE", "Operation #$index: Flag muss 0 oder 1 sein.")
                        MatriboxFamilyExpansionCatalog.Kind.NUMBER ->
                            if (op.value < 0f) reject("VALUE_TYPE", "Operation #$index: negativer Wert für einen Zahlenparameter.")
                        else -> Unit
                    }
                    if (!written.add(slot to op.name)) reject("CONFLICTING_OPERATIONS", "Operation #$index: ${op.name} doppelt in ${slot.name}.")
                }
                is FullLiveOperation.BlockToggle ->
                    if (!toggled.add(slot)) reject("CONFLICTING_OPERATIONS", "Operation #$index: ${slot.name} wird doppelt geschaltet.")
            }
            // the closed codec must be able to encode it (no store/metadata family exists there)
            try {
                MatriboxFullLiveCodec.validate(MatriboxFullLiveCodec.encode(op))
            } catch (error: IllegalArgumentException) {
                reject("ENCODING_INVALID", "Operation #$index: ${error.message}")
            }
        }
    }
}
